src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/DiagnosticDescriptorExtensions.cs

```diff
@@ -55,6 +56,44 @@ public static ImmutableArray<string> ImmutableCustomTags(this DiagnosticDescript
             return effectiveSeverity;
         }
 
+        /// <summary>
+        /// Gets document-level effective severity of the given <paramref name="descriptor"/> accounting for severity configurations from both the following sources:
+        /// 1. Compilation options from ruleset file, if any, and command line options such as /nowarn, /warnaserror, etc.
+        /// 2. Analyzer config documents at the document root directory or in ancestor directories.
+        /// </summary>
+        public static ReportDiagnostic GetEffectiveSeverity(this DiagnosticDescriptor descriptor, CompilationOptions compilationOptions, SyntaxTree tree, AnalyzerOptions analyzerOptions)
+        {
+            var effectiveSeverity = descriptor.GetEffectiveSeverity(compilationOptions);
+
+            // Apply analyzer config options, unless configured with a non-default value in compilation options.
+            // Note that compilation options (/nowarn, /warnaserror) override analyzer config options.
+            if (!compilationOptions.SpecificDiagnosticOptions.TryGetValue(descriptor.Id, out var reportDiagnostic) ||
+                reportDiagnostic == ReportDiagnostic.Default)
```

Comment:
❔ Do document-level configurations not have priority over compilation-level?

Reply:
No, this follows the precedence order used by the compiler: See first bullet point in https://learn.microsoft.com/dotnet/fundamentals/code-analysis/configuration-files#severity-options and also see https://github.com/dotnet/roslyn/blob/0b9e4d612bc4d02ee8b3ead3be95b2f2a6bd1e93/src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs#L115-L129